src/python/deepgnn/graph_engine/snark/decoders.py

```diff
@@ -43,6 +43,191 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        default_edge_feature_types: ["dtype" or None, ...] Dtype of each feature vector.
+        default_edge_feature_lens: [[int, ...] or None, ...] Length value for each feature vector.
+
+    Init Parameters Format Example
+        Same graph as the previous example, just with defaults specified.
+        ```
+        EdgeListDecoder(
+            default_node_type=1,
+            default_node_weight=.5,
+            default_node_feature_types=["int32", "float32"],
+            default_node_feature_lens=[[3],[2]],
+            default_edge_type=0,
+            default_edge_weight=.5,
+            default_edge_feature_types=["uint8"],
+                default_edge_feature_lens=[[3, 0]],
+        )
+        ```
+        graph.csv
+        ```
+        0 -1 1 1 1 1.1 1.1
+        0 1 0 4 10 1 1 1
+        1 -1 1 1 1 1.1 1.1
+        1 0 0 4 10 1 1 1
+        ```
+    """
+
+    def __init__(
+        self,
+        default_node_type: Optional[int] = None,
+        default_edge_type: Optional[int] = None,
+        default_node_weight: Optional[float] = None,
+        default_edge_weight: Optional[float] = None,
+        default_node_feature_types: Optional[List[Optional[str]]] = None,
+        default_edge_feature_types: Optional[List[Optional[str]]] = None,
+        default_node_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        default_edge_feature_lens: Optional[List[Optional[List[int]]]] = None,
+    ):
+        """Initialize the Decoder."""
+        super().__init__()
+        self.node_type = default_node_type
+        self.node_weight = default_node_weight
+        self.node_feature_types = (
+            default_node_feature_types if default_node_feature_types is not None else []
+        )
+        self.node_feature_lens = (
+            default_node_feature_lens if default_node_feature_lens is not None else []
+        )
+        self.n_node_feature = len(self.node_feature_types)
+
+        self.edge_type = default_edge_type
+        self.edge_weight = default_edge_weight
+        self.edge_feature_types = (
+            default_edge_feature_types if default_edge_feature_types is not None else []
+        )
+        self.edge_feature_lens = (
+            default_edge_feature_lens if default_edge_feature_lens is not None else []
+        )
+        self.n_edge_feature = len(self.edge_feature_types)
+
+    def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        """Convert text line into a node and edge iterator.
+
+        This is a generator that yields a node then its outgoing edges in order.
+        A node and its outgoing edges can be on different partitions.
+        Yield format is (node_id/src, -1/dst, type, weight, features).
+        Features being a list of dense features as ndarrays and sparse features as 2 tuples, coordinates and values.
+        """
+        if line == "":
+            return []
+
+        data = iter(line.split(","))
```

Comment:
line.split returns a list of strings, which uses lots of memory. Why can't we use a CSV reader?

Reply:
csv reader also goes line by line: 
https://github.com/python/cpython/blob/da4aae29f829fdf288b42919671274d7d463fe04/Lib/csv.py#:~:text=self.fieldnames-,row%20%3D%20next(self.reader),-self.line_num

Each line is only a single node or edge so the limit is very large feature vectors

Reply:
I am talking about parsing lines, not reading them from files. 
The difference between str.split and CSV reader is the latter reader doesn't create new strings, but [updates C fields](https://github.com/python/cpython/blob/f62ad4f2c4214fdc05cc45c27a5c068553c7942c/Modules/_csv.c#L657) which is way faster.

Reply:
Each line has a different length and we can't know the types beforehand which csv reader does not support.

Reply:
Of course csv reader supports variable length and different types. Using example from the docs:
![image](https://user-images.githubusercontent.com/134559/183712713-ab6ebea4-446a-4575-9fc6-6c5ed1ebed3e.png)

Reply:
Yes but we cant take advantage of the benefits of the csv reader with the different length rows. This looks like string.split with extra overhead?

Reply:
1. What overhead are you talking about?
2. Add a test case with a node binary feature equal to string: "parsing, not splitting". I don't think existing parser can handle it.

Reply:
1. I benchmarked, CSV reader is faster even for just string splitting
2. I dont think either soltion will support this, I'll parameterize the split operators

Reply:
Actually I forgot to do next(csv.reader), it is actually 25-50% slower than .split